### PR TITLE
inital commit: use offical elastic JS client, refactor ability to spe…

### PR DESCRIPTION
**Pull request by yomashExpel**
…cify bulk options and to be more generic.

### CODE REVIEW

File: examples/elastic_basic.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    // If the parser fails event.data will contain `@type` set to 'unparseable' and `parseError` will be
+    // the reason the parser failed
+    // If the parser succeeds the original message will be backed up in event.data.originalMessage
+    // The third argument controls whether or not the original message is backed up,
+    // set it to false or don't provide it to disable the feature
+    if (!parsers.jsonParser(event, '@type', true)) {
+        return event.done()
+    }
+
+    var _type = 'foo',
```
Comment (yomashExpel) [inline examples/elastic_basic.js]: specify bulk actions here versus in the Elasticoutput, provides more flexibility without having to map various options to the Elasticoutput.

File: examples/elastic_basic.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    // parsers is a global object containing all parsers built into StreamStash
+    // If the parser fails event.data will contain `@type` set to 'unparseable' and `parseError` will be
+    // the reason the parser failed
+    // If the parser succeeds the original message will be backed up in event.data.originalMessage
+    // The third argument controls whether or not the original message is backed up,
+    // set it to false or don't provide it to disable the feature
+    if (!parsers.jsonParser(event, '@type', true)) {
+        return event.done()
+    }
+
```
Comment (yomashExpel) [inline examples/elastic_basic.js]: specify bulk actions here versus in the Elasticoutput, provides more flexibility without having to map various options to the Elasticoutput.

**Comment by nbrownus** [comment]
Thanks for the PR! I'll have to look into what the official ES client offers, specifically I want to make sure we can easily configure how many sockets the underlying http agent will allow concurrently as that could impact performance. I'll comment on some specific lines for more suggestions.

File: lib/outputs/ElasticSearchOutput.js [hunk 3]
```diff
@@ -2,111 +2,79 @@ var util = require('util'),
...
 
-    if (useOptions.auth) {
-        self.httpOptions.auth = useOptions.auth
-    }
+    self.options = useOptions.options  || {};
 
     self.queueOptions = {
         batchSize: useOptions.batchSize || 100,
         timeout: useOptions.batchTimeout || 1000
-    }
```
Comment (nbrownus) [inline lib/outputs/ElasticSearchOutput.js]: `;` is entirely optional and should be avoided unless the situation specifically calls for it

File: lib/outputs/ElasticSearchOutput.js [hunk 4]
```diff
@@ -2,111 +2,79 @@ var util = require('util'),
...
+    };
```
Comment (nbrownus) [inline lib/outputs/ElasticSearchOutput.js]: ; is entirely optional and should be avoided unless the situation specifically calls for it

File: lib/outputs/ElasticSearchOutput.js [hunk 5]
```diff
@@ -163,98 +131,49 @@ ElasticSearchOutput.prototype._handleOutput = function (eventContainer) {
...
+        var event = queue.items[eventId];
+        bulkBody.push(event.data.bulkActionMeta);
+        // TODO if index / update?
+        bulkBody.push(event.data.message);
     }
 
-    if (write === '') {
-        // Don't bother sending emtpy requests
-        return
+    if(_.size(bulkBody) > 0) {
```
Comment (nbrownus) [inline lib/outputs/ElasticSearchOutput.js]: `bulkBody.length` should be faster than underscores `.size`

File: lib/outputs/ElasticSearchOutput.js [hunk 6]
```diff
@@ -163,98 +131,49 @@ ElasticSearchOutput.prototype._handleOutput = function (eventContainer) {
...
+        self.elasticClient.bulk(_.extend(self.options, {body: bulkBody})).then(function(response) {
```
Comment (nbrownus) [inline lib/outputs/ElasticSearchOutput.js]: `util._extend()` would work as well and then you could avoid importing underscore entirely.

File: package.json [hunk 7]
```diff
@@ -1,44 +1,45 @@
...
-        "relp": "git://github.com/nbrownus/node-relp",
-        "statsd-client": "0.2.1"
-    },
-    "devDependencies": {
-        "istanbul": "0.4.x",
-        "ppunit": "x",
-        "should": "7.1.x",
-        "jsdoc": "3.3.3"
-    }
+  "name": "streamstash",
```
Comment (nbrownus) [inline package.json]: Try to avoid unnecessary white space changes

**Pull request closed without merging**